### PR TITLE
Add `pkgs.lib.encodeGNUCommandLine`

### PR DESCRIPTION
**Pull request by Gabriella439**
###### Motivation for this change

This adds a new utility to intelligently convert Nix records to
command line options to reduce boilerplate for simple use cases and to
also reduce the likelihood of malformed command lines

###### Things done

I ran `nix-instantiate --eval --strict lib/tests/misc.nix`

### CODE REVIEW

**Comment by edolstra** [comment]
I'm hesitant about this because it puts a highly specific policy of how command line arguments are encoded in the standard library. It's not at all clear how universally applicable this is. For example, `v = false` is encoded as an empty string, but maybe in some CLI tools you want it to be encoded as `--no-v`.

The name `renderOptions` is rather vague. At first I thought this had something to do with the NixOS module system. Maybe something like `encodeGNUCommandLine` is more descriptive.

Escaping shell arguments should be handled elsewhere since you could also pass arguments as an array. So escaping could be done separately (e.g. `concatMap escapeShellArg (renderOptions ...)`).

**Comment by Gabriel439** [comment]
@edolstra: I would be open to naming this in such a way that it does not imply that it is *the* only way to render command-line options and/or making the utility's behavior customizable (e.g. customizing how it handles lists or boolean-valued fields).  For example, conceptually this utility could take an additional record argument specifying how to handle each behavior (with default values so that the user can supply `{ }` for the common case):

```nix
encodeGNUCommandLine =
    { renderBool ? key: value: if value then " ${hyphenate key}" else ""
    , renderList ? key: values: ...
    , renderOption ? key: value: ...
    , ...
    }
    options: ...
```

**Comment by Gabriel439** [comment]
@edolstra: Alright, I believe this is ready for re-review now.  I incorporated your feedback

File: lib/tests/misc.nix [hunk 1]
```diff
@@ -441,4 +441,21 @@ runTests {
...
+         { foo = "A";
+
+           bar = 1;
+
+           baz = null;
+
+           qux = true;
+
+           v = true;
+         };
```
Comment (roberth) [inline lib/tests/misc.nix]: This should also have list examples and a `false` value.

**Comment by Gabriel439** [comment]
@roberth: I believe I addressed your feedback and this is ready for another review

**Comment by jtojnar** [comment]
Where would this be used seeing it returns string? We are moving towards `__structuredAttrs`, where things like `configureFlags` will be lists: https://github.com/NixOS/nixpkgs/pull/72074

**Comment by Gabriel439** [comment]
@jtojnar: A `systemd` service specification is one example of where you can use this:

```nix
{ config, lib, pkgs, ... }:

let
  cfg = config.services.example;

  options = { inherit (cfg) host port; };

in
  { options.services.example = {
      enable = lib.mkEnableOption "example";

      host = lib.mkOption { ... };

      port = lib.mkOption { ... };
    };

    config.systemd.services.example = lib.mkIf cfg.enable {
      wantedBy = [ "multi-user.target" ];

      script = ''
        ${pkgs}/bin/example${lib.renderOptions options}
      '';
    };
  }
```

**Comment by roberth** [comment]
You could extract a function `toGNUCommandLine` from this: `encodeGNUCommandLine = a: escapeShellArgs (toGNUCommandLine a)`

`toGNUCommandLine` will then be usable with structured attrs.

This will require a preceding space as in `the-cmd ${encodeGNUCommandLine attrs}`, but that's good because it's consistent with `escapeShellArgs` and the shell itself.

**Comment by jtojnar** [comment]
Well, the function returns a string. To be useful for structured attrs, it would have to return a list.

**Comment by roberth** [comment]
> return a list

precisely

File: lib/default.nix [hunk 2]
```diff
@@ -120,6 +121,7 @@ let
       isOptionType mkOptionType;
     inherit (asserts)
       assertMsg assertOneOf;
+    inherit (cli) encodeGNUCommandLine;
```
Comment (roberth) [inline lib/default.nix]:
```suggestion
    inherit (cli) encodeGNUCommandLine toGNUCommandLine;
```

**Review by roberth — APPROVED** [review]

**Review by infinisil — APPROVED** [review]
I like it, this could be used to provide something like NixOS/rfcs#42 for CLI configs.

**Comment by Profpatsch** [comment]
Just seeing this, it would have been good if it had followed the same interface as `lib/generators.nix`;
that is naming the options `mkX` instead of `renderX`. The `fn {} val` interface is the same, luckily.

Not entirely sure what the separation in `renderList/Option/Bool` is for, one argument with pattern match would suffice I think. As long as it’s new like this, can we still change the interface?

**Comment by Profpatsch** [comment]
Oh, I’d also rename the functions to use `toX` for the string output (which is how generators.nix names its functions) and `toXList` for the list version.

**Comment by Profpatsch** [comment]
Ah, no, the string is explicitely POSIX shell-formatted, so we should call it that. `toXShell`. See `escapeShellArg` in `strings.nix.`

**Comment by Profpatsch** [comment]
Let’s also not expose it into the global namespace, similar to `generators`. `cli` is only three letters.

**Comment by roberth** [comment]
@Profpatsch feel free to make a PR for the rename/move.

> Not entirely sure what the separation in renderList/Option/Bool is for

It's for overriding the convention, for example to allow `x = false` to be rendered as `--no-x`.

Re POSIX: it does seem to be GNU, at least according to GNU.

> GNU adds long options to these conventions. Long options consist of ‘--’ followed by a name made of alphanumeric characters and dashes.

-- https://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html

I'd welcome a more generic term than GNU, but I don't know of any that implies long options.

**Comment by Profpatsch** [comment]
I created a PR with my suggested breaking changes: https://github.com/NixOS/nixpkgs/pull/78337

**Pull request merged**